File: src/ScpiLab/Measurement.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Text;

namespace ScpiLab
{
    public sealed class SweepPoint
    {
        public double SetVoltage;
        public double MeasuredVoltage;
        public double MeasuredCurrent;
    }

    public sealed class SweepResult
    {
        public List<SweepPoint> Points = new List<SweepPoint>();

        public double MeanCurrent()
        {
            if (Points.Count == 0) return 0.0;
            double sum = 0.0;
            foreach (var p in Points) sum += p.MeasuredCurrent;
            return sum / Points.Count;
        }

        public double StdDevCurrent()
        {
            if (Points.Count < 2) return 0.0;
            double mean = MeanCurrent(), acc = 0.0;
            foreach (var p in Points) { var d = p.MeasuredCurrent - mean; acc += d * d; }
            return Math.Sqrt(acc / (Points.Count - 1));
        }

        public double MaxVoltageError()
        {
            double worst = 0.0;
            foreach (var p in Points)
            {
                var e = Math.Abs(p.MeasuredVoltage - p.SetVoltage);
                if (e > worst) worst = e;
            }
            return worst;
        }

        /// <summary>导出 CSV（测试数据要能留痕、能被别的工具再分析）。</summary>
        public string ToCsv()
        {
            var sb = new StringBuilder();
            sb.AppendLine("set_voltage_V,measured_voltage_V,measured_current_A");
            foreach (var p in Points)
            {
                sb.AppendLine(string.Format(CultureInfo.InvariantCulture, "{0},{1},{2}",
                    p.SetVoltage, p.MeasuredVoltage, p.MeasuredCurrent));
            }
            return sb.ToString();
        }
    }

    /// <summary>可复用的测量脚本：电压扫描 + 采集 + 统计（对应"测试方法标准化、可复用"）。</summary>
    public static class SweepRunner
    {
        public static SweepResult RunVoltageSweep(PowerSupply psu, double from, double to, double step, int settleReads = 3)
        {
            if (psu == null) throw new ArgumentNullException("psu");
            if (step <= 0.0 || to < from) throw new ScpiException("扫描参数不合法");

            var result = new SweepResult();
            for (double v = from; v <= to + 1e-9; v += step)
            {
                psu.SetVoltage(v);
                double measuredV = 0.0, measuredI = 0.0;
                for (int i = 0; i < Math.Max(1, settleReads); i++)
                {
                    measuredV = psu.MeasureVoltage();
                    measuredI = psu.MeasureCurrent();
                }
                result.Points.Add(new SweepPoint
                {
                    SetVoltage = v,
                    MeasuredVoltage = measuredV,
                    MeasuredCurrent = measuredI
                });
            }
            return result;
        }
    }
}

File: scpi_lab/measurement.py
"""可复用的测量脚本：电压扫描 → 采集 → 统计 → CSV 导出（对应"测试方法标准化、可复用"）。"""
from __future__ import annotations

import math
import statistics
from dataclasses import dataclass, field
from typing import List

from .instruments import PowerSupply
from .session import ScpiError


@dataclass
class SweepPoint:
    set_voltage: float
    measured_voltage: float
    measured_current: float


@dataclass
class SweepResult:
    points: List[SweepPoint] = field(default_factory=list)

    def mean_current(self) -> float:
        return statistics.fmean([p.measured_current for p in self.points]) if self.points else 0.0

    def std_current(self) -> float:
        if len(self.points) < 2:
            return 0.0
        return statistics.stdev([p.measured_current for p in self.points])

    def max_voltage_error(self) -> float:
        return max((abs(p.measured_voltage - p.set_voltage) for p in self.points), default=0.0)

    def to_csv(self) -> str:
        lines = ["set_voltage_V,measured_voltage_V,measured_current_A"]
        for p in self.points:
            lines.append("%.6g,%.6g,%.6g" % (p.set_voltage, p.measured_voltage, p.measured_current))
        return "\n".join(lines) + "\n"


class SweepRunner:
    """电压扫描：每个点设置后连读 settle_reads 次，取最后一次作为稳定读数。"""

    @staticmethod
    def run_voltage_sweep(psu: PowerSupply, start: float, stop: float, step: float,
                          settle_reads: int = 3) -> SweepResult:
        if step <= 0.0 or stop < start:
            raise ScpiError("扫描参数不合法：start=%s stop=%s step=%s" % (start, stop, step))
        result = SweepResult()
        count = int(math.floor((stop - start) / step + 1e-9)) + 1
        for i in range(count):
            volts = start + i * step
            psu.set_voltage(volts)
            measured_v = measured_i = 0.0
            for _ in range(max(1, settle_reads)):
                measured_v = psu.measure_voltage()
                measured_i = psu.measure_current()
            result.points.append(SweepPoint(volts, measured_v, measured_i))
        return result
